design/open/kv3-execution-plan.md

```diff
@@ -0,0 +1,52 @@
+
+# General
+## Tables
+Graveler uses 5 DB tables:
+* `graveler_repositories`,`graveler_branches`,`graveler_commits`,`graveler_tags` - used by `pkg/graveler/ref`
+* `graveler_staging_kv` - used by `pkg/graveler/staging`
+
+`graveler_repositories`, `graveler_commits` and `graveler_tags` are immutable and does not update entries, and so should be straight forward to implement over our `KV` (`graveler_commits` actually performs update for commit generation, but this is only valid for `refs-restore` which can only take place on a `bare` repo and so it is of no concern)</br>
+`graveler_branches` is currently being protected by a lock, during commit actions, and in order to support lockless commits with `KV`, the [following design](https://github.com/treeverse/lakeFS/blob/b3204edad00f88f8eb98524ad940fde96e02ab0a/design/open/metadata_kv/index.md#graveler-metadata---branches-and-staged-writes) is proposed. However, as a first step we can implement it using a naive table lock, just to get it going, or better, a `branch-key` based locking, to prevent staging during commits</br>
+`graveler_staging_kv` creates entries based on the current branch `staging_token` and is protected from creating entries during a commit under the same lock. This should also be supported lockless-ly, as referred to by the above proposal, and can also be solved with naive locking, as a first step</br>
+Note: All the above tables (along with the late `auth_users`) are also used by `pkg/diagnostics` for diagnostics collection (RO). 
+
+## Iterators
+The following iterators are implemented and used by `pkg/graveler`:
+  * `branch_iterator` - iterates branches by repository. Has an option to order by ID (trivial) or by CommitID (Used by GC. How do we support that?)
+  * `commit_iterator` - iterates a commit ancestry by running BFS. Not an `iterator` in the `KV` sense. Should be supported as is with `KV` `Get`
+  * `commit_ordered_iterator` - iterates commits of a repo, ordered by commit ID (regular `KV` `Scan` supports that). However - has an option to iterate over not-first-parent commits only (e.g. commits that are not a first parent of any other commit). Probably could (and should) be implemented by a specific `Next()` for that case
```

Comment:
Needs more details. The list of `not-first-parent commits` can always change since you may introduce a new child to any commit at any time. I wonder how do we use this list? And why would the ordering of CommitID matter?

I find it hard to come up with an iterator on top of the kv implementation that would cover these requirements.

Reply:
These commits are used side by side with the `branch-by-commitID` list, as explained by @johnnyaug, hence the order by `commitID` for that list too.
regarding the `not-first-parent` attribute, its usage and the fact it is subtle to changes - is it any different than the way it is used today, over `SQL`? I see that the decision of whether a commit is `non-first-parent` or not is made at the time of query execution, but this also can be changed by the time the GC algorithm get there. Am I wrong?
I guess I miss some context on that algorithm - @johnnyaug, can you please help here?

Reply:
> I find it hard to come up with an iterator on top of the kv implementation that would cover these requirements.

@itaiad200 - I absolutely agree with that last sentence. Maybe we can maintain a live set of `not-first-parent` commits and modify it as commits occur. i.e. each new commit is by default `not-first-parent` and for each further operation, which makes it a `first-parent`, we **try** to remove that commit from the list (possibly already removed, hence the **try**) (**Question**: Can a commit **become** `not-first-parent` if it is already a `first-parent`?)
I can think of a couple of ways to  implement this list, each optnio with its own drawbacks:
* We can use some `not_first_parent/<COMMIT_ID>`. This will require require searching the list for each commit, or
* We can "flag" the commit key itself i.e. commit `commit/<COMMIT_ID>` is accompanied by `commit/<COMMIT_ID>/not_first_parent` (not sure this works well with our scan logic, right?)
What other options do we have?

Reply:
> In the end, this doesn't need to be a consistent list: when you find a commit that is not a first parent, you can safely return it, regardless of whether the resulting list as a whole makes sense.

Thank you for this input @johnnyaug.
What happens if I found a `not-first-parent` commit, return it as a starting point, and before the GC contniues, a commit is created over that commit, making it a `first parent`? Is that possible?

Reply:
I think that we might be overworking this.  The fact that something is efficiently possible on PostgreSQL does not mean that we have to do it efficiently in the KV world and in the same way.

Also please note that the proposal to add parameters to `Next()` simply makes it _not an iterator_.  I would argue that the iterator abstraction adds nothing here.

We will not be able to manage this list of "not first-parent commits" correctly and efficiently.  The question is really, do we need to?  We will probably be scanning all commits, or all commits up to a previously-processed set of "safe" commits, on each GC.  We could process not-first commits "just" by scanning all of these.  (I am starting to think that the commits logic for GC should _also_ move off of lakeFS and into the Spark cluster.  @johnnyaug WDYT?)

Reply:
I'm totally open to it - the not-first-parent thing is a means to an end. We just need to start scanning the tree somewhere.

Reply:
I'm aiming at preserving the current behavior with KV support, as a first step, and improve that later, based on our needs. I think introducing the KV aspect is enough of a load already, and would like to stay away from redesigning our GC as a side effect.
Even though it will not be as efficient, I think it can work, at least as a first step. Moreover, since it is the GC, I think we can loosen up the efficiency requirements, again - at least as a first step.
@arielshaqed @johnnyaug - WDYT?

Reply:
@arielshaqed - either ways we are not changing the interface of `Next()`

Reply:
@itai-david I understood that we will be reading everything to memory, applying the same logic that was used in Postgres, thus implementing the same iterators as today. If so, what are the remaining open questions?